src/Test/StateMachine/Parallel.hs

```diff
@@ -315,48 +327,99 @@ runParallelCommandsNTimes :: (Show (cmd Concrete), Show (resp Concrete))
+                           -> PropertyM m [(History cmd resp, Logic)]
+runParallelCommandsNTimes' n sm complete cmds =
+  replicateM n $ do
+    (hist, _reason1, _reason2) <- run (executeParallelCommands sm cmds)
+    let hist' = completeHistory complete hist
+    return (hist', linearise sm hist')
+
 executeParallelCommands :: (Show (cmd Concrete), Show (resp Concrete))
                         => (Rank2.Traversable cmd, Rank2.Foldable resp)
                         => (MonadCatch m, MonadUnliftIO m)
                         => StateMachine model cmd m resp
                         -> ParallelCommands cmd resp
-                        -> m (History cmd resp, Reason)
+                        -> m (History cmd resp, Reason, Reason)
 executeParallelCommands sm@StateMachine{ initModel } (ParallelCommands prefix suffixes) = do
 
   hchan <- newTChanIO
 
   (reason0, (env0, _smodel, _counter, _cmodel)) <- runStateT
-    (executeCommands sm hchan (Pid 0) True prefix)
+    (executeCommands sm hchan (Pid 0) CheckEverything prefix)
     (emptyEnvironment, initModel, newCounter, initModel)
 
   if reason0 /= Ok
   then do
     hist <- getChanContents hchan
-    return (History hist, reason0)
+    return (History hist, reason0, reason0)
   else do
-    (reason, _) <- foldM (go hchan) (reason0, env0) suffixes
+    (reason1, reason2, _) <- go hchan (Ok, Ok, env0) suffixes
     hist <- getChanContents hchan
-    return (History hist, reason)
+    return (History hist, reason1, reason2)
   where
-    go hchan (_, env) (Pair cmds1 cmds2) = do
+    go _hchan (res1, res2, env) []                         = return (res1, res2, env)
+    go hchan  (Ok,   Ok,   env) (Pair cmds1 cmds2 : pairs) = do
+
       ((reason1, (env1, _, _, _)), (reason2, (env2, _, _, _))) <- concurrently
 
         -- XXX: Post-conditions not checked, so we can pass in initModel here...
-        -- It would be better if we made executeCommands take a Maybe model
-        -- instead of the boolean...
-
-        (runStateT (executeCommands sm hchan (Pid 1) False cmds1) (env, initModel, newCounter, initModel))
-        (runStateT (executeCommands sm hchan (Pid 2) False cmds2) (env, initModel, newCounter, initModel))
-      return ( reason1 `combineReason` reason2
-             , env1 <> env2
-             )
-      where
-        combineReason :: Reason -> Reason -> Reason
-        combineReason Ok r2 = r2
-        combineReason r1 _  = r1
+        -- It would be better if we made executeCommands take a Maybe Environment
+        -- instead of the Check...
+
+        (runStateT (executeCommands sm hchan (Pid 1) CheckNothing cmds1) (env, initModel, newCounter, initModel))
+        (runStateT (executeCommands sm hchan (Pid 2) CheckNothing cmds2) (env, initModel, newCounter, initModel))
+      go hchan ( reason1
+               , reason2
+               , env1 <> env2
+               ) pairs
+    go hchan (Ok, ExceptionThrown, env) (Pair cmds1 _cmds2 : pairs) = do
```

Comment:
@stevana is there any reason to continue execution if there is an ExceptionThrown?

Reply:
Sequential execution also stops if there is an exception.

Reply:
I think we should stop execution for a specific process (or `Pid`) if there's an `ExceptionThrown` on that process (but continue executing on the other processes). I believe that's what I've implemented here.

This is also why we need your genernalisation to allow execution on more than two processes, so that we can introduce faults (exceptions) without stopping all execution altogether.

Reply:
I see, I guess it depends on how we want to simulate user usage of the api and since this gives more user option I think it's on the right direction. @stevana do you think this is close to being merged? I see some errors like `executeCommands: impossible` and some false positives tests, which I believe will be solved with this pr.

Reply:
Another important thing that this PR adds is the ability to _complete_ a history which contains crashed processes. We do so by appending a user specified response to the end of the history. For example in the case of the memory reference example, if a `Write` fails then we can have a history like this:

```
Process 1: |-- x <- Create --|                           |---- Read x ---|
Process 2:                       |---- Write 5 x  --...        
```
In this case `Read` can return both `0` (the default value of create) or `5` depending on if `Write` crashed before or after writing.

To account of this we  complete the above history as follows:
```
Process 1: |-- x <- Create --|                           |---- Read x ---|
Process 2:                       |---- Write 5 x  -------------------------|    
```

Notice that the `Write` is concurrent with the `Read`. Now because of the way `linearise` works (it tries all possible sequential interleavings of the action calls) it will accept `Read` returning both `0` and `5`.

To complete for `Write` is easy as it's just an `Ack`, and it doesn't really matter what we complete `Read` with as it doesn't change the model. If a `Create` crashes, we are kind of screwed that's why in this PR I also allowed the pre-condition failed exception to be thrown. No doubt will there be examples for which complete will be harder...

A possible alternative would be to have the user catch all exceptions (due to fault injection) and account of the non-determinism, e.g. have not just a single value for each memory reference, but a set of values. This complicates the model and is a lot more work for the user though.

Yet another possibility might be to have the fault injection be more precise, so that we know if the `Write` failed before or after it wrote to the memory. That way we know if it should return `0` or `5`. I'm not sure if having this precision is always possible. It also complicates the model because we need to keep track of exactly what faults are injected and return the correct response in the presence of the fault.

Does this make sense? If not, let me know where I lost you and I can try to explain further.

It's my current understanding of the Linearizability paper and what Jepsen does. It took me a while to get here, and I'm still not certain if I understand things correctly. So I think it would be great if: 1) I could convince you that this approach makes sense, and 2) that we develop more examples that confirm that it indeed works.